File: app/src/main/java/com/soumikshah/investmenttracker/view/MainActivity.java
package com.soumikshah.investmenttracker.view;

import android.app.DatePickerDialog;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.Toolbar;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.soumikshah.investmenttracker.R;
import com.soumikshah.investmenttracker.database.model.Investment;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;


public class MainActivity extends AppCompatActivity {

    public CoordinatorLayout getCoordinatorLayout() {
        return coordinatorLayout;
    }

    private CoordinatorLayout coordinatorLayout;

    private ViewPager viewPager;
    private BottomNavigationView bottomNavigationView;
    MainFragment mainFragment;
    long investmentDateInLong = 0;
    DatePickerDialog datePickerDialog;
    float interestToBeReceived;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setActionBar(toolbar);
        if(getActionBar() !=null){
            getActionBar().setDisplayHomeAsUpEnabled(false);
            getActionBar().hide();
        }
        mainFragment = new MainFragment();
        viewPager = findViewById(R.id.viewpager);
        bottomNavigationView = findViewById(R.id.bottom_navigation);
        setupViewPager(viewPager);

        bottomNavigationView.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch(item.getItemId()){
                    case R.id.mainPage:
                        viewPager.setCurrentItem(0);
                        break;
                    case R.id.graph:
                        viewPager.setCurrentItem(1);
                        break;
                    case R.id.settings:
                        viewPager.setCurrentItem(2);
                        break;
                    default:
                        throw new IllegalStateException("Unexpected value: " + item.getItemId());
                }
                return true;
            }
        });

        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                switch (position) {
                    case 0:
                        bottomNavigationView.getMenu().findItem(R.id.mainPage).setChecked(true);
                        break;
                    case 1:
                        bottomNavigationView.getMenu().findItem(R.id.graph).setChecked(true);
                        break;
                    case 2:
                        bottomNavigationView.getMenu().findItem(R.id.settings).setChecked(true);
                        break;
                }
            }
            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });

        coordinatorLayout = findViewById(R.id.coordinator_layout);
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                showInvestmentDialog(false, null, -1);
            }
        });
    }

    private void setupViewPager(ViewPager viewPager){
        ViewPagerAdapter adapter = new ViewPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(mainFragment, "Mainpage");
        adapter.addFragment(new GraphFragment(), "Graph");
        adapter.addFragment(new SettingsFragment(),"Settings");
        viewPager.setAdapter(adapter);
    }

    static class ViewPagerAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        public ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        public void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitleList.get(position);
        }
    }
    /**
     * Shows alert dialog with EditText options to enter / edit
     * a note.
     * when shouldUpdate=true, it automatically displays old note and changes the
     * button text to UPDATE
     */
    void showInvestmentDialog(final boolean shouldUpdate, final Investment investment, final int position) {
        LayoutInflater layoutInflaterAndroid = LayoutInflater.from(MainActivity.this);
        View view = layoutInflaterAndroid.inflate(R.layout.investment_dialog, null);

        AlertDialog.Builder alertDialogBuilderUserInput = new AlertDialog.Builder(MainActivity.this);
        alertDialogBuilderUserInput.setView(view);

        final EditText inputInvestmentName = view.findViewById(R.id.investment);
        final EditText inputInvestmentAmount = view.findViewById(R.id.investmentAmount);
        final EditText inputInvestmentPercent = view.findViewById(R.id.investmentInterest);
        final EditText inputInvestmentMedium = view.findViewById(R.id.investmentMedium);
        final EditText inputInvestmentCategory = view.findViewById(R.id.investmentCategory);
        final TextView inputInvestmentDate = view.findViewById(R.id.investedDate);
        final EditText inputInvestmentNumberOfMonths = view.findViewById(R.id.investedNumberOfMonths);


        TextView dialogTitle = view.findViewById(R.id.dialog_title);
        dialogTitle.setText(!shouldUpdate ? getString(R.string.new_investment_title) : getString(R.string.edit_investment_title));

        inputInvestmentDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // calender class's instance and get current date , month and year from calender
                final Calendar c = Calendar.getInstance();
                int mYear = c.get(Calendar.YEAR); // current year
                int mMonth = c.get(Calendar.MONTH); // current month
                int mDay = c.get(Calendar.DAY_OF_MONTH); // current day
                // date picker dialog
                datePickerDialog = new DatePickerDialog(MainActivity.this,R.style.DatePickerTheme,
                        new DatePickerDialog.OnDateSetListener() {

                            @Override
                            public void onDateSet(DatePicker view, int year,
                                                  int monthOfYear, int dayOfMonth) {
                                // set day of month , month and year value in the edit text
                                inputInvestmentDate.setText(String.format(Locale.ENGLISH,"%d/%d/%d", dayOfMonth, monthOfYear + 1, year));
                                investmentDateInLong = c.getTimeInMillis();
                            }
                        }, mYear, mMonth, mDay);
                datePickerDialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.DKGRAY));
                datePickerDialog.show();
            }
        });

        if (shouldUpdate && investment != null) {
            inputInvestmentName.setText(investment.getInvestmentName());
            inputInvestmentAmount.setText(String.valueOf(investment.getInvestmentAmount()));
            inputInvestmentPercent.setText(String.valueOf(investment.getInvestmentPercent()));
            inputInvestmentMedium.setText(String.valueOf(investment.getInvestmentMedium()));
            inputInvestmentCategory.setText(String.valueOf(investment.getInvestmentCategory()));
            SimpleDateFormat sim = new SimpleDateFormat("dd/MM/YYYY",Locale.ENGLISH);
            inputInvestmentDate.setText(sim.format(investment.getInvestmentDate()));
            inputInvestmentNumberOfMonths.setText(String.valueOf(investment.getInvestmentMonth()));
        }
        alertDialogBuilderUserInput
                .setCancelable(false)
                .setPositiveButton(shouldUpdate ? "update" : "save", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialogBox, int id) {

                    }
                })
                .setNegativeButton("cancel",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialogBox, int id) {
                                dialogBox.cancel();
                            }
                        });

        final AlertDialog alertDialog = alertDialogBuilderUserInput.create();
        alertDialog.setOnKeyListener(new DialogInterface.OnKeyListener() {
            @Override
            public boolean onKey(DialogInterface dialogInterface, int i, KeyEvent keyEvent) {
                if (keyEvent.getKeyCode() == KeyEvent.KEYCODE_BACK) {
                    alertDialog.dismiss();
                }
                return false;
            }
        });
        alertDialog.show();

        alertDialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Show toast message when no text is entered
                if (TextUtils.isEmpty(inputInvestmentName.getText().toString())) {
                    Toast.makeText(MainActivity.this, "Enter Investment Name!", Toast.LENGTH_SHORT).show();
                    return;
                } else if (TextUtils.isEmpty(inputInvestmentAmount.getText().toString())){
                    Toast.makeText(MainActivity.this,"Enter Investment Amount!", Toast.LENGTH_LONG).show();
                    return;
                } else if(TextUtils.isEmpty(inputInvestmentCategory.getText().toString())){
                    Toast.makeText(MainActivity.this,"Enter Investment Type!",Toast.LENGTH_LONG).show();
                }else {
                    alertDialog.dismiss();
                }

                if(inputInvestmentPercent.getText().toString().matches("")){
                    interestToBeReceived = 0;
                }else{
                    interestToBeReceived = Float.parseFloat(inputInvestmentPercent.getText().toString());
                }

                if(inputInvestmentMedium.getText().toString().isEmpty()){
                    inputInvestmentMedium.setText(R.string.not_mentioned);
                }
                if(inputInvestmentCategory.getText().toString().isEmpty()){
                    inputInvestmentCategory.setText("");
                }
                if(inputInvestmentNumberOfMonths.getText().toString().isEmpty()){
                    inputInvestmentNumberOfMonths.setText("0");
                }

                if (shouldUpdate && investment != null) {
                    mainFragment.investmentHelper.updateInvestment(inputInvestmentName.getText().toString(),
                            Integer.parseInt(inputInvestmentAmount.getText().toString()),
                            interestToBeReceived,
                            inputInvestmentMedium.getText().toString(),
                            inputInvestmentCategory.getText().toString(),
                            investmentDateInLong,
                            Integer.parseInt(inputInvestmentNumberOfMonths.getText().toString()),
                            position);
                } else {
                    mainFragment.investmentHelper.createInvestment(inputInvestmentName.getText().toString(),
                            Integer.parseInt(inputInvestmentAmount.getText().toString()),
                            interestToBeReceived,
                            inputInvestmentMedium.getText().toString(),
                            inputInvestmentCategory.getText().toString(),
                            investmentDateInLong,Integer.parseInt(inputInvestmentNumberOfMonths.getText().toString()));
                }
            }
        });
    }


}

File: app/src/main/java/com/soumikshah/investmenttracker/view/MainActivity.kt
package com.soumikshah.investmenttracker.view

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.View.GONE
import android.widget.Toast
import android.widget.Toolbar
import androidx.appcompat.app.AppCompatActivity
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.soumikshah.investmenttracker.R
import com.soumikshah.investmenttracker.view.widget.TotalAmountInvestedWidget
import java.util.*
import java.util.concurrent.Executor


class MainActivity : AppCompatActivity() {
    var coordinatorLayout: CoordinatorLayout? = null
        private set
    private var viewPager: ViewPager? = null
    private var bottomNavigationView: BottomNavigationView? = null
    @JvmField
    var mainFragment: MainFragment? = null
    private var fab: FloatingActionButton? =null
    private var adapter:ViewPagerAdapter? = null

    override fun onDestroy() {
        super.onDestroy()
        //Update the widget when application is closed, so any changes made will be reflected in the widget right away.
        val intent = Intent(this,TotalAmountInvestedWidget::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        val ids = AppWidgetManager.getInstance(application)
            .getAppWidgetIds(ComponentName(applicationContext,
                TotalAmountInvestedWidget::class.java))
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        sendBroadcast(intent)
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        coordinatorLayout = findViewById(R.id.coordinator_layout)
        fab = findViewById<View>(R.id.fab) as FloatingActionButton
        val executor = ContextCompat.getMainExecutor(this)
        val biometricManager = BiometricManager.from(this)
        when (biometricManager.canAuthenticate()) {
            BiometricManager.BIOMETRIC_SUCCESS ->
            {
                coordinatorLayout!!.visibility = GONE
                authUser(executor)
            }
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE ->
                Toast.makeText(
                    this,
                    getString(R.string.error_msg_no_biometric_hardware),
                    Toast.LENGTH_LONG
                ).show()
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE ->
                Toast.makeText(
                    this,
                    getString(R.string.error_msg_biometric_hw_unavailable),
                    Toast.LENGTH_LONG
                ).show()
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED ->
                Toast.makeText(
                    this,
                    getString(R.string.error_msg_biometric_not_setup),
                    Toast.LENGTH_LONG
                ).show()
        }

        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setActionBar(toolbar)
        if (actionBar != null) {
            actionBar!!.setDisplayHomeAsUpEnabled(false)
            actionBar!!.hide()
        }
        mainFragment = MainFragment()
        viewPager = findViewById(R.id.viewpager)
        bottomNavigationView = findViewById(R.id.bottom_navigation)
        setupViewPager(viewPager)
        bottomNavigationView!!.setOnNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.mainPage -> viewPager!!.currentItem = 0
                R.id.settings -> {
                    viewPager!!.currentItem = 1
                }
                /*R.id.graph -> viewPager!!.currentItem = 1*/
                else -> {
                    throw IllegalStateException("Unexpected value: " + item.itemId)
                }
            }
            true
        }

        viewPager!!.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }
            override fun onPageSelected(position: Int) {
                when (position) {
                    0 -> {
                        bottomNavigationView!!.menu.findItem(R.id.mainPage).isChecked = true
                        viewPager!!.adapter!!.notifyDataSetChanged()
                    }
                    1 -> {
                        bottomNavigationView!!.menu.findItem(R.id.settings).isChecked = true
                        viewPager!!.adapter!!.notifyDataSetChanged()
                    }
                }
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })

        fab!!.setOnClickListener {
            loadFragment(ShowDialog(false, null, -1))
        }
    }

    fun showFab(){
        fab!!.show()
    }
    fun hideFab(){
        fab!!.hide()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        val currentFragment =
            this.supportFragmentManager.findFragmentById(R.id.mainPage)
    }
    fun loadFragment(someFragment: Fragment){
        val fragment: Fragment = someFragment
        val transaction = supportFragmentManager.beginTransaction()
        transaction.setCustomAnimations(R.anim.slide_in_right,R.anim.slide_out_left,
            R.anim.slide_in_left,R.anim.slide_out_right)
        transaction.add(R.id.fragment, fragment)
        transaction.addToBackStack(null) // if written, this transaction will be added to backstack
        transaction.commit()
    }

    fun updateViewPager(){
        if(adapter!=null){
            adapter!!.notifyDataSetChanged()
        }
    }
    private fun setupViewPager(viewPager: ViewPager?) {
        adapter = ViewPagerAdapter(supportFragmentManager)
        adapter!!.addFragment(mainFragment, getString(R.string.mainpage_viewpager_name))
        adapter!!.addFragment(SettingsFragment(), getString(R.string.setttings_viewpager_name))
        adapter.also { viewPager!!.adapter = it }
    }

    private fun authUser(executor: Executor) {
        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(getString(R.string.auth_title))
            .setSubtitle(getString(R.string.auth_subtitle))
            .setDescription(getString(R.string.auth_description))
            .setDeviceCredentialAllowed(true)
            .build()

        val biometricPrompt = BiometricPrompt(this, executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(
                    result: BiometricPrompt.AuthenticationResult
                ) {
                    super.onAuthenticationSucceeded(result)
                    coordinatorLayout!!.visibility = View.VISIBLE
                }
                override fun onAuthenticationError(
                    errorCode: Int, errString: CharSequence
                ) {
                    super.onAuthenticationError(errorCode, errString)
                    Toast.makeText(applicationContext,
                        errString.toString(),
                        Toast.LENGTH_SHORT).show()
                }
                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    Toast.makeText(applicationContext,
                        getString(R.string.error_msg_auth_failed),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
        biometricPrompt.authenticate(promptInfo)
    }
    class ViewPagerAdapter(manager: FragmentManager?) : FragmentPagerAdapter(manager!!) {
        private val mFragmentList: MutableList<Fragment> = ArrayList()
        private val mFragmentTitleList: MutableList<String> = ArrayList()
        override fun getItem(position: Int): Fragment {
            return mFragmentList[position]
        }

        override fun getCount(): Int {
            return mFragmentList.size
        }

        fun addFragment(fragment: Fragment?, title: String) {
            mFragmentList.add(fragment!!)
            mFragmentTitleList.add(title)
        }

        override fun getPageTitle(position: Int): CharSequence {
            return mFragmentTitleList[position]
        }

        override fun getItemPosition(`object`: Any): Int {
            return POSITION_NONE
        }
    }
}